### ISSUE TITLE
GUI redesign phase 1

### ISSUE THREAD

**Issue by HEdingfield** [BrightSpots/rcv#461]
Implement the simplest changes specified in the Center for Civic Design Universal Tabulator Redesign Final mockups document (`Tabulator (FINAL mockups) May 2020.docx`).

These will mainly consist of moving existing options to new tabs and rearranging the order they appear in.

**Comment by HEdingfield** [comment]
Provider dropdown should change type of dialog box opened up by "select" button for CVR file path (e.g. CDF should only allow JSON, ES&S should only allow .xlsx and .xls, Hart should only allow directory, etc.).

**Comment by HEdingfield** [comment]
Refugees from #152:

* Delete Selected buttons should be disabled if nothing is selected
* Ability to drag and edit rows (e.g. for candidate permutations)

**Comment by HEdingfield** [comment]
Notes after reviewing the final UI mockups doc today:



@hedingfield:
* Some stuff was moved to the Output tab from Contest Info, and I'm definitely not going to restructure the config files at all in light of this, so our .txt description doc will have things out of order which isn't ideal in the short-term unless I can get to adding the hints
* In this phase, I won't be doing the overvote label, treat blank as UWI settings, etc. on a per-source-file level now since that'd necessitate a config structure change
* I've got general concerns about fitting too much stuff in a panel due to resolution issues... should I bump it to 1600x1024 as the minimum required resolution, or even bigger? And if a display doesn't have that much space, I'll need to auto-expand it and figure out how to make scrollbars behave properly.
* I'm also not a huge fan of disappearing the log window, and it likely won't happen in this phase.

@tarheel:
* For winner election mode, they recommended eliminating support for "single seat until 2 candidates remain." But we need to support that somehow. I guess the alternative is to go back to what I believe we had before, which is to make "continue until two remain" a boolean in the config. We could put it under the output settings, since, as they point out, it only affects how the output is presented, not who actually wins.
* When the mode is bottoms up with percentage, number of winners should not be a field that the user sets. (Right now we have a single multiSeatBottomsUp mode, and we differentiate between percentage-based bottoms-up vs. the normal kind based on whether the config sets numberOfWinners to zero or sets a value for multiSeatBottomsUpPercentageThreshold.)

@moldover:
* Contest Info:
  * The hint: "fields do not influence computations..." isn't accurate as Dominion and Hart filter by contest ID, Clear Ballot filters by Contest Name, etc..
* CVR files:
  * Hint text needs re-working here to account for different providers
* Candidates:Candidate File:
  * For Dominion pre-populating candidates could happen fully automatically - there is a canonical candidates location and file.
  * For Clear Vote, there is a candidate file you could specify.
  * Other vendors I'm not sure about.
* Consider not splitting up contest info from output (not sure why this is desired)
* Don't hesitate to bump up resolution to avoid dynamic scroll-bars.  We are not supporting mobile, and I can't imagine a laptop which doesn't have 1024 at least.
* Hide/Show log does seem useful, OR, if the error pop-up gets to a good place, we really don't need to show the log at all.  It would mostly be part of the output files and for debugging / power users.

**Comment by HEdingfield** [comment]
Per the latest in #456, `contestId` will become a per-CVR setting.

In addition, @moldover suggested:
> Maybe update Contest Name prompt in UI to (something painfully obvious like) 'Output Report Contest Name' and then Contest ID becomes: 'Contest ID to Tabulate'.

**Comment by chughes297** [comment]
Saw this issue when looking at the contest ID thing so wanted to add some info on a couple comments:

> For winner election mode, they recommended eliminating support for "single seat until 2 candidates remain." But we need to support that somehow. I guess the alternative is to go back to what I believe we had before, which is to make "continue until two remain" a boolean in the config. We could put it under the output settings, since, as they point out, it only affects how the output is presented, not who actually wins.

Sorry just noticed that this is unclear in the mockups! We found in testing that admins couldn't parse the difference between running down to two and stopping at a majority when explained in text, for some reason, so there's only one single-winner election mode. But there's a checkbox for running all the way down to two in the winning rules mockup on page 18 in the Final Mockups doc. Does that work? (also just noticed that checkbox only appears on page 18 and not in other single-winner mockups, so that probably made it even more confusing).

> Consider not splitting up contest info from output (not sure why this is desired)

In testing we found that people were thrown off when the first tab asked them about where output files should go. Users thought of the contest info tab as just starting up with the contest in question, and file location/other output info was a different frame of mind for them.

**Comment by HEdingfield** [comment]
Thanks for the feedback, Chris, will take all this into account. May need to file a separate issue for your first point because it could entail a non-backwards-compatible change to the config file, which is definitely tough.

**Comment by chughes297** [comment]
Ok. Seems like there's a few proposed revisions that will cause non-backwards-compatibility, so are those going in a later phase? 

Also, George and I have discussed this briefly and we aren't too worried about losing backwards compatibility. Is there something in particular about losing it that makes you all uneasy?

**Comment by tarheel** [comment]
I agree that making non-backward-compatible changes shouldn't be a huge deal at this stage, especially if the tabulator shows a helpful error message instead of just crashing.

But also, for some changes, we can just write simple logic to migrate an old config to the new version behind the scenes, so it will be basically transparent to the user. The proposed "single seat until 2 candidates remain" change is a good example of that.

**Comment by HEdingfield** [comment]
From @chughes297 regarding the hints for the Voter Error and Output tabs (which were missing from the mockup:

Here's what I've drafted. Lynn and I didn't take the time to write these back during testing so this hasn't been put in front of election admins directly, but I've written down how I described these options to admins as we were testing the UI. 

(the voter error hints include examples/demonstrations of how each rule works, but worry this will make the hint box too large - I'm putting examples that could feasibly be deleted for brevity in brackets)

---

[Voter Error Tab](https://a89u6z.axshare.com/#g=1&p=voter_error_rules)

The tabulator needs to know how to handle voter errors in your jurisdiction. These requirements are typically included in statute or regulation. 

**Overvote Rule** (required): how to handle a ballot where a voter has marked multiple candidates at the same ranking when that ballot is encountered in the round-by-round count. This rule relies on overvote labels as filled out on the CVR tab.

&nbsp;&nbsp;&nbsp;&nbsp;[**Always skip to the next rank** skips over an overvote and goes to the next validly marked ranking on a ballot.
&nbsp;&nbsp;&nbsp;&nbsp;**Exhaust immediately** means that a ballot with an overvote exhausts when that overvote is encountered in the rounds of counting.
&nbsp;&nbsp;&nbsp;&nbsp;**Exhaust if Multiple Continuing** - if a voter has an overvote but only one candidate at that overvote is still in the race when that overvote is encountered, the ballot counts for that candidate. If multiple candidates at the overvote are still in the race, the ballot exhausts.]

**How Many Consecutive Skipped Ranks Are Allowed** (required): how many rankings in a row can a voter skip and still have later rankings count? 0 allows no skipped rankings. 1 allows voters to skip rankings one at a time, but not more than 1 in a row, and so on. [Example: A voter could rank in 1, 3, 5 and not exhaust under this rule, for example. Implementation of this rule depends on the undervote labels as filled out on the CVR tab.]

**Exhaust on Duplicate Ranking**: When checked, the tabulator will exhaust a ballot that includes multiple rankings for the same candidate when that repeat ranking is reached. [Example: A voter ranks the same candidate 1st and 3rd, a different candidate 2nd, and another candidate 4th. If their original first choice and their second choice are eliminated, the ballot exhausts when it reaches the repeat ranking in rank 3. The ranking in the 4th rank does not count.]

[Output tab](https://a89u6z.axshare.com/#g=1&p=output)

Tell the tabulator where results files go and what additional results files you want.

**Output directory**: The location where results files will go.

**Prefix for output file names**: A unique label for results files so you can more easily tell different tabulations apart. Files are named by date and time by default. 

**Tabulate by Precinct**: Produce round-by-round results at the precinct level. Results are how ballots at the precinct level transferred in the contest as a whole, not a simulated round-by-round count in the precinct. Requires precinct information in CVR tab.

**Generate a CVR-level output file**: Produce an additional file that shows what happens to every ballot in every round (this might be a very large file)

**Generate a CDF JSON**: Produce a VVSG common data format JSON file of the CVR.

**Comment by HEdingfield** [comment]
Notes on Winner Election Mode from email conversation with @chughes297:

* "Single-winner majority determines winner"
    * Maps to "standard" with `numberOfWinners` set to 1
* "Multi-winner allow only one winner per round"
    * Maps to "multiSeatAllowOnlyOneWinnerPerRound"
* "Multi-winner allow multiple winners per round"
    * Maps to "standard" with the user required to provide `numberOfWinners` >1
* "Bottoms up"
    * Maps to "multiSeatBottomsUp"
* "Bottoms up using percentage threshold"
    * Maps to "multiSeatBottomsUp" with `numberOfWinners` set to 0 and the user required to provide `multiSeatBottomsUpPercentageThreshold` > 0
* "Multi-pass IRV"
    * Maps to "multiSeatSequentialWinnerTakesAll"

Another note:
* "singleSeatContinueUntilTwoCandidatesRemain"
    * No longer in the Winner Election Mode menu; instead this is set in the config if the user selects "Single-winner majority determines winner" and checks a new checkbox in the GUI for "Continue until 2 candidates remain"
